src/scenechange/mod.rs

```diff
@@ -241,64 +223,119 @@ impl<T: Pixel> SceneChangeDetector<T> {
+      self.cost_scenecut(frame1, frame2)
     };
-    self
-      .score_deque
-      .insert(0, (result.inter_cost as f64, result.threshold as f64));
+
+    // Subtract the highest metric value of surrounding frames from the current one
+    // It makes the peaks in the metric more distinct
+    if self.speed_mode != SceneDetectionSpeed::Fast && self.deque_offset > 0 {
+      if input_frameno == 1 {
+        // Accounts for the second frame not having a score to adjust against.
+        // It should always be 0 because the first frame of the video is always a keyframe.
+        result.backward_adjusted_cost = 0.0;
+      } else {
+        let mut adjusted_cost = f64::MAX;
+        for other_cost in
+          self.score_deque.iter().take(self.deque_offset).map(|i| i.inter_cost)
+        {
+          let this_cost = result.inter_cost - other_cost;
+          if this_cost < adjusted_cost {
+            adjusted_cost = this_cost;
+          }
+          if adjusted_cost < 0.0 {
+            adjusted_cost = 0.0;
+            break;
+          }
+        }
+        result.backward_adjusted_cost = adjusted_cost;
+      }
+      if !self.score_deque.is_empty() {
+        for i in 0..(cmp::min(self.deque_offset, self.score_deque.len())) {
+          let adjusted_cost =
+            self.score_deque[i].inter_cost - result.inter_cost;
+          if i == 0
+            || adjusted_cost < self.score_deque[i].forward_adjusted_cost
+          {
+            self.score_deque[i].forward_adjusted_cost = adjusted_cost;
+          }
+          if self.score_deque[i].forward_adjusted_cost < 0.0 {
+            self.score_deque[i].forward_adjusted_cost = 0.0;
+          }
+        }
+      }
+    }
+    self.score_deque.insert(0, result);
   }
 
   /// Compares current scene score to adapted threshold based on previous scores
   /// Value of current frame is offset by lookahead, if lookahead >=5
   /// Returns true if current scene score is higher than adapted threshold
-  fn adaptive_scenecut(&mut self) -> bool {
-    // Subtract the previous metric value from the current one
-    // It makes the peaks in the metric more distinctive
-    if (self.speed_mode != SceneDetectionSpeed::Fast) && self.deque_offset > 0
+  fn adaptive_scenecut(&mut self) -> (bool, ScenecutResult) {
+    let score = self.score_deque[self.deque_offset];
+
+    // We use the histogram algorithm's cost metrics as a secondary algorithm
+    // because, although it struggles in certain scenarios such as
+    // finding the end of a pan, it is very good at detecting hard scenecuts
+    // or detecting if a pan exists.
+    // Because of this, we only consider a frame for a scenechange if
+    // the histogram algorithm is over the threshold either on this frame (hard scenecut)
+    // or within the past few frames (pan). This helps filter out a few false positives
+    // produced by the cost-based algorithm.
+    let hg_threshold = HISTOGRAM_THRESHOLD * (self.bit_depth as f64) / 8.0;
```

Comment:
We can do this separately, but we should probably rename the histogram method.

Reply:
Done